File: apps/desktop/src/MaroonedSoftware.Deadair.Desktop.Core/Net/StationHttp.cs
namespace MaroonedSoftware.Deadair.Desktop.Core.Net;

/// <summary>
/// The one <see cref="HttpClient"/> everything in the app goes through.
/// </summary>
/// <remarks>
/// <para>
/// <b>One client, not one per concern.</b> The station counts an HLS listener by IP and User-Agent,
/// so an app whose API calls, artwork fetches and audio all present differently is counted as
/// several listeners, and one that sends no agent at all is counted as whatever the platform's
/// default is that release. Putting the agent on a handler here means a new caller cannot forget it.
/// </para>
/// <para>
/// The SDK is handed this client and told not to dispose it, and is given NO header callback of its
/// own: the bearer goes on in a handler too, so that the refresh and its single replay live in one
/// place rather than beside every call.
/// </para>
/// </remarks>
public static class StationHttp
{
    /// <summary>
    /// Long enough for a station under load and short enough that a wrong address fails while somebody
    /// is still looking at the screen.
    /// </summary>
    private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(15);

    /// <summary>
    /// Builds the client.
    /// </summary>
    /// <param name="session">
    /// The handler that attaches the bearer and performs the single-flight refresh, when there is
    /// one. Null before sign-in exists, which is every phase up to the one that adds it.
    /// </param>
    public static HttpClient Create(DelegatingHandler? session = null)
    {
        var transport = new SocketsHttpHandler
        {
            // The station is one host and the app talks to it constantly. Pooling is what keeps a
            // two-second poll from being a TLS handshake every two seconds.
            PooledConnectionLifetime = TimeSpan.FromMinutes(10),
            ConnectTimeout = TimeSpan.FromSeconds(10),
            AutomaticDecompression = System.Net.DecompressionMethods.All,
        };

        DelegatingHandler outermost = new UserAgentHandler();
        DelegatingHandler requestId = new RequestIdHandler();

        outermost.InnerHandler = requestId;

        if (session is not null)
        {
            requestId.InnerHandler = session;
            session.InnerHandler = transport;
        }
        else
        {
            requestId.InnerHandler = transport;
        }

        return new HttpClient(outermost, disposeHandler: true)
        {
            Timeout = RequestTimeout,
        };
    }
}

File: apps/ios/Sources/Net/StationHttp.swift
import DeadairCore
import DeadairSdk
import Foundation

/// The app's one HTTP stack: every API call and every piece of artwork goes through `session`.
///
/// One session because the station counts an HLS listener per IP and User-Agent, so the agent has to
/// be the same on every request the app makes. It is set twice, deliberately: on the session, which
/// is what the artwork loader relies on, and by `AgentTransport` on every SDK request, which
/// overwrites whatever a caller set. The player cannot use a session at all and gets the same string
/// through `AVURLAssetHTTPUserAgentKey`.
final class StationHttp: Sendable {
    let userAgent: String
    let session: URLSession

    init(userAgent: String) {
        self.userAgent = userAgent
        let configuration = URLSessionConfiguration.default
        configuration.httpAdditionalHeaders = ["User-Agent": userAgent]
        // Fifteen seconds without a byte, which is what the other two apps allow a request: long
        // enough for a station waking up, short enough that "could not reach" arrives while
        // somebody is still looking at the screen.
        configuration.timeoutIntervalForRequest = 15
        configuration.waitsForConnectivity = false
        session = URLSession(configuration: configuration)
    }

    /// A client for one station, carrying whatever headers it is handed on every request.
    func sdk(for station: StationUrl, headers: @escaping @Sendable () async throws -> [String: String] = { [:] }) -> Deadair {
        Deadair(config: SdkConfig(
            baseURL: station.apiBaseURL,
            headers: headers,
            transport: AgentTransport(URLSessionTransport(session: session), userAgent: userAgent)
        ))
    }

    /// The agent this build sends: the app's name and its marketing version from the bundle.
    static var bundleAgent: String {
        DeadairCore.userAgent(version: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String)
    }
}
